aips/aip-20.md

```diff
@@ -0,0 +1,464 @@
+
+    /// Compute a pre-compiled pairing function (a.k.a., bilinear map) on `element_1` and `element_2`.
+    /// Return an element in the target group `Gt`.
+    public fun pairing<G1,G2,Gt>(element_1: &Element<G1>, element_2: &Element<G2>): Element<Gt>;
+
+    /// Try deserializing a byte array to an element of an algebraic structure `S` using a given serialization format `F`.
+    /// Return none if the deserialization failed.
+    public fun deserialize<S, F>(bytes: &vector<u8>): Option<Element<S>>;
+
+    /// Serialize an element of an algebraic structure `S` to a byte array using a given serialization format `F`.
+    public fun serialize<S, F>(element: &Element<S>): vector<u8>;
+
+    /// Get the order of structure `S`, a big integer little-endian encoded as a byte array.
+    public fun order<G>(): vector<u8>;
+
+    /// Cast an element of a structure `S` to a super-structure `L`.
+    public fun upcast<S,L>(element: &Element<S>): Element<L>;
+
+    /// Try casting an element `x` of a structure `L` to a sub-structure `S`.
+    /// Return none if `x` is not a member of `S`.
+    ///
+    /// NOTE: Membership check is performed inside, which can be expensive, depending on the structures `L` and `S`.
+    public fun downcast<L,S>(element_x: &Element<L>): Option<Element<S>>;
+
+    /// Hash an arbitrary-length byte array `msg` into structure `S` with a domain separation tag `dst`
+    /// using the given hash-to-structure suite `H`.
+    public fun hash_to<St, Su>(dst: &vector<u8>, msg: &vector<u8>): Element<St>;
+
+    #[test_only]
+    /// Generate a random element of an algebraic structure `S`.
+    public fun rand_insecure<S>(): Element<S>;
+
+```
+
+In general, every structure implements basic operations like (de)serialization, equality check, random sampling.
+
+A group may also implement the following operations. (Additive notions are used.)
+- `order()` for getting the group order.
+- `zero()` for getting the group identity.
+- `one()` for getting the group generator (if exists).
+- `neg()` for group element inversion.
+- `add()` for basic group operation.
+- `sub()` for group element subtraction.
+- `double()` for efficient group element doubling.
+- `scalar_mul()` for group scalar multiplication.
+- `multi_scalar_mul()` for efficient group multi-scalar multiplication.
+- `hash_to()` for hash-to-group.
+
+A field may also implement the following operations.
+- `order()` for getting the field order.
+- `zero()` for the field additive identity.
+- `one()` for the field multiplicative identity.
+- `add()` for field addition.
+- `sub()` for field subtraction.
+- `mul()` for field multiplication.
+- `div()` for field division.
+- `neg()` for field negation.
+- `inv()` for field inversion.
+- `sqr()` for efficient field element squaring.
+- `from_u64()` for quick conversion from u64 to field element.
+
+For 3 groups that form a bilinear map, `pairing()` and `multi_pairing()` may be implemented.
```

Comment:
Editorial: "Similarly, for 3 groups `G1`, `G2`, `GT` that **admit** a bilinear map [...], `pairing<G1, G2, GT>()` and [...]."

Reply:
Fixed.